In [163]:
import pandas as pd
import numpy as np
import re
import time

import bs4 as bs4
import json

import glob
import tqdm

pd.set_option("max.columns", 131)

#https://strftime.org/
%matplotlib inline
%pylab inline

Populating the interactive namespace from numpy and matplotlib


/Users/mario/miniconda3/lib/python3.7/site-packages/IPython/core/magics/pylab.py:160: UserWarning: pylab import has clobbered these variables: ['vstack', 'hstack']
`%matplotlib` prevents importing * from pylab and numpy
  "\n`%matplotlib` prevents importing * from pylab and numpy"


# Resultados do Active Learning

In [164]:
df1 = pd.read_csv("raw_data_with_labels.csv", index_col=0)
df1 = df1[df1['y'].notnull()]
df1.shape

(498, 16)

In [165]:
df2 = pd.read_csv("active_labels1_done.csv", index_col=0)
df2 = df2[df2['y'].notnull()]
df2['novo'] = 1
df2.shape

(100, 18)

In [166]:
df2.head(1)

watch-title  y     watch-view-count  \
505  Platform Overview - Machine Learning  0  4.298 visualizações   

                     watch-time-text content_watch-info-tag-list  \
505  Publicado em 21 de mai. de 2019        Ciência e tecnologia   

                          watch7-headline  \
505  Platform Overview - Machine Learning   

                                    watch7-user-header  \
505  Google Cloud Platform\n\n\n\n\n\n\n\n\n\n\n\n\...   

                              watch8-sentiment-actions  \
505  4.298 visualizações\n\n\n\n\n\n\n\n141\n\nGost...   

                                              og:image  og:image:width  \
505  https://i.ytimg.com/vi/QR_LQQ-vvko/maxresdefau...          1280.0   

     og:image:height                                     og:description  \
505            720.0  In this short GCP Essentials video, see how GC...   

     og:video:width  og:video:height               og:video:tag  \
505          1280.0            720.0  Alexis Moussine Pouchkine   

                        channel_link_0      p  novo  
505  /channel/UCJS9pqu9BzkAMNTmzNMNhvg  0.502     1

In [167]:
from sklearn.metrics import roc_auc_score, average_precision_score
average_precision_score(df2['y'],df2['p']), roc_auc_score(df2['y'],df2['p'])

(0.2037344613689981, 0.5386250885896527)

In [ ]:
df = pd.concat([df1, df2.drop("p", axis=1)])

In [141]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [142]:
df_limpo = pd.DataFrame(index=df.index)
df_limpo['title'] = df['watch-title']
df_limpo['novo'] = df['novo'].fillna(0)

## 1. Limpeza da data

In [168]:
clean_date = df['watch-time-text'].str.extract(r"(\d+) de ([a-z]+)\. de (\d+)")
clean_date[0] = clean_date[0].map(lambda x: "0"+x[0] if len(x) == 1 else x)
#clean_date[1] = clean_date[1].map(lambda x: x[0].upper()+x[1:])

mapa_meses = {"jan": "Jan",
              "fev": "Feb",
              "mar": "Mar", 
              "abr": "Apr", 
              "mai": "May", 
              "jun": "Jun",
              "jul": "Jul",
              "ago": "Aug", 
              "set": "Sep", 
              "out": "Oct", 
              "nov": "Nov",
              "dez": "Dec"}

clean_date[1] = clean_date[1].map(mapa_meses)

clean_date = clean_date.apply(lambda x: " ".join(x), axis=1)
clean_date.head()
df_limpo['date'] = pd.to_datetime(clean_date, format="%d %b %Y")

## 2. Limpeza de Views

In [169]:
views = df['watch-view-count'].str.extract(r"(\d+\.?\d*)", expand=False).str.replace(".", "").fillna(0).astype(int)
df_limpo['views'] = views

## 3. Features

In [170]:
features = pd.DataFrame(index=df_limpo.index)
y = df['y'].copy()

In [171]:
features['tempo_desde_pub'] = (pd.to_datetime("2019-12-03") - df_limpo['date']) / np.timedelta64(1, 'D')
features['views'] = df_limpo['views']
features['views_por_dia'] = features['views'] / features['tempo_desde_pub']
features = features.drop(['tempo_desde_pub'], axis=1)

In [172]:
features.head()

views  views_por_dia
0  28028      61.464912
1   1131       2.960733
2   1816       8.446512
3   1171      10.455357
4   1228       3.336957

# Aumenta validação

In [162]:
mask_train = (df_limpo['date'] < "2019-04-01") & (df_limpo['novo'] == 0)

mask_val = (df_limpo['date'] >= "2019-04-01")

Xtrain, Xval = features[mask_train], features[mask_val]
ytrain, yval = y[mask_train], y[mask_val]
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

((228, 2), (316, 2), (228,), (316,))

In [173]:
from sklearn.feature_extraction.text import TfidfVectorizer

title_train = df_limpo[mask_train]['title']
title_val = df_limpo[mask_val]['title']

title_vec = TfidfVectorizer(min_df=2)
title_bow_train = title_vec.fit_transform(title_train)
title_bow_val = title_vec.transform(title_val)


In [174]:
title_bow_train.shape

(228, 193)

In [175]:
from scipy.sparse import hstack, vstack

Xtrain_wtitle = hstack([Xtrain, title_bow_train])
Xval_wtitle = hstack([Xval, title_bow_val])

In [176]:
Xtrain_wtitle.shape, Xval_wtitle.shape

((228, 195), (316, 195))

In [177]:
mdl = RandomForestClassifier(n_estimators=1000, random_state=0, class_weight="balanced", n_jobs=6)
mdl.fit(Xtrain_wtitle, ytrain)

RandomForestClassifier(bootstrap=True, class_weight='balanced',
            criterion='gini', max_depth=None, max_features='auto',
            max_leaf_nodes=None, min_impurity_decrease=0.0,
            min_impurity_split=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            n_estimators=1000, n_jobs=6, oob_score=False, random_state=0,
            verbose=0, warm_start=False)

In [178]:
p = mdl.predict_proba(Xval_wtitle)[:, 1]

In [179]:
from sklearn.metrics import roc_auc_score, average_precision_score

In [180]:
average_precision_score(yval, p), roc_auc_score(yval, p)

(0.1872802752830275, 0.5935436218282933)

ANTES  
ap 0.1918043901336543, auc 0.5848024316109421 - mindf=2  


# Aumenta treino

In [183]:
mask_train = (df_limpo['date'] < "2019-04-01")
#mask_val = (df_limpo['date'] >= "2019-04-01") & (df_limpo['novo'] == 0)
mask_val = (df_limpo['date'] >= "2019-04-01")

In [184]:
Xtrain, Xval = features[mask_train], features[mask_val]
ytrain, yval = y[mask_train], y[mask_val]
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

title_train = df_limpo[mask_train]['title']
title_val = df_limpo[mask_val]['title']

title_vec = TfidfVectorizer(min_df=2)
title_bow_train = title_vec.fit_transform(title_train)
title_bow_val = title_vec.transform(title_val)

Xtrain_wtitle = hstack([Xtrain, title_bow_train])
Xval_wtitle = hstack([Xval, title_bow_val])


mdl = RandomForestClassifier(n_estimators=1000, random_state=0, class_weight="balanced", n_jobs=6)
mdl.fit(Xtrain_wtitle, ytrain)

p = mdl.predict_proba(Xval_wtitle)[:, 1]

average_precision_score(yval, p), roc_auc_score(yval, p)

(0.19342838733336654, 0.6095324991310394)

ANTES  
ap 0.1918043901336543, auc 0.5848024316109421 - mindf=2  

Valid nova, treino antigo  
ap 0.1872802752830275, auc 0.5935436218282933

Valid antiga, treino novo  
ap 0.17697903398635972, auc 0.5964133738601823

Tudo novo  
ap 0.19342838733336654, auc 0.6095324991310394